### ISSUE TITLE
sources: a postgres source with incorrect authentication parameters produces bad error messages

### ISSUE THREAD

**Issue by chaas** [MaterializeInc/materialize#20098]
From @SangJunBak:

> When a pg connection fails because of incorrect authentication credentials, we just get
> `ERROR: error connecting to server: Connection timed out (os error 110): Connection timed out (os error 110)`.
> 
> Is it possible to get a more descriptive hint/error similar to above?
> 
> I noticed for things like incorrect egress IPs/network security, we get a hint back about changing firewall settings and for an incorrect broker, we get `ERROR: error connecting to server: failed to lookup address information: Name or service not known: failed to lookup address information: Name or service not known` which is awesome.

([Slack](https://materializeinc.slack.com/archives/C02FWJ94HME/p1687280519230829?thread_ts=1686911558.162349&cid=C02FWJ94HME))

Similar to https://github.com/MaterializeInc/materialize/issues/12465, which is about kafka error messages.

**Comment by petrosagg** [comment]
@SangJunBak  I don't see the same behavior, and in fact we have tests that verify that the error message is more descriptive here https://github.com/petrosagg/materialize/blob/connection-purification/test/pg-cdc/pg-cdc.td#L176-L191

Are you sure you got the connection details right in your test?

**Comment by dseisun-materialize** [comment]
@SangJunBak same as Petros, I tried to repro and got the more specific error message when using a bad username or password. Do you happen to remember the setup that was giving you the unspecific error message?

**Comment by SangJunBak** [comment]
Ah I figured it out. Sorry for the delay and confusion! Four things:

1. The reason that error was happening was because the db instance OOMed. If there's a way to detect that, would be helpful as an error message or notice/hint! The combination I tried was a correct broker, correct database name, incorrect username. 
2. When I reset the instance and made sure it had enough memory, I got this error: 
<img width="300" alt="Screenshot 2023-07-06 at 4 39 15 PM" src="https://github.com/MaterializeInc/materialize/assets/29664023/393c5748-9480-4294-95d0-85fe26c611ca">

which is a lot more descriptive! However, I noticed the error message is repeated. Is this a bug or intended?
3. Same when you give it an improper username/password combo:
<img width="400" alt="Screenshot 2023-07-06 at 4 40 16 PM" src="https://github.com/MaterializeInc/materialize/assets/29664023/721ca8fb-084c-41c4-b30b-e98b1aee64c0">

4. I noticed if you give it a non-existent database name, it'll succeed. Is this intended?
